run/idp-sql/database.py

```diff
@@ -0,0 +1,201 @@
+                )  # i.e "<PROJECT-NAME>:<INSTANCE-REGION>:<INSTANCE-NAME>"
+            },
+        ),
+        **db_config,
+    )
+    pool.dialect.description_encoding = None
+    logger.info("Database engine initialised from unix conection")
+
+    return pool
+
+
+# [END cloudrun_user_auth_sql_connect]
+
+
+def create_tables() -> None:
+    # This is called before any request on the main app, ensuring the database has been setup
+    global db
+    db = init_connection_engine()
+    # Create pet_votes table if it doesn't already exist
+    with db.connect() as conn:
+        conn.execute(
+            "CREATE TABLE IF NOT EXISTS pet_votes"
+            "( vote_id SERIAL NOT NULL, "
+            "time_cast timestamp NOT NULL, "
+            "candidate VARCHAR(6) NOT NULL, "
+            "uid VARCHAR(128) NOT NULL, "
+            "PRIMARY KEY (vote_id)"
+            ");"
+        )
+
+
+def get_index_context() -> Dict:
+    votes = []
+    with db.connect() as conn:
+        # Execute the query and fetch all results
+        recent_votes = conn.execute(
+            "SELECT candidate, time_cast FROM pet_votes "
+            "ORDER BY time_cast DESC LIMIT 5"
+        ).fetchall()
+        # Convert the results into a list of dicts representing votes
+        for row in recent_votes:
+            votes.append(
+                {
+                    "candidate": row[0],
+                    "time_cast": row[1],
+                }
+            )
+        stmt = sqlalchemy.text(
+            "SELECT COUNT(vote_id) FROM pet_votes WHERE candidate=:candidate"
+        )
+        # Count number of votes for cats
+        cats_result = conn.execute(stmt, candidate="CATS").fetchone()
+        cats_count = cats_result[0]
+        # Count number of votes for dogs
+        dogs_result = conn.execute(stmt, candidate="DOGS").fetchone()
+        dogs_count = dogs_result[0]
+
+    return {
+        "dogs_count": dogs_count,
+        "recent_votes": votes,
+        "cats_count": cats_count,
+    }
+
+
+def save_vote(team: str, uid: str, time_cast: datetime.datetime) -> None:
+    # Preparing a statement before hand can help protect against injections.
+    stmt = sqlalchemy.text(
+        "INSERT INTO pet_votes (time_cast, candidate, uid)"
+        " VALUES (:time_cast, :candidate, :uid)"
+    )
+
+    # Using a with statement ensures that the connection is always released
+    # back into the pool at the end of statement (even if an error occurs)
+    with db.connect() as conn:
+        conn.execute(stmt, time_cast=time_cast, candidate=team, uid=uid)
+
+    logger.info("Vote for %s saved.", team)
+
+
+def shutdown() -> None:
+    # Each connection was released on execution, so just formally
+    # dispose of the db connection if it's been instantiated
+    if db:
+        db.dispose()
+        logger.info("Database connection disposed.")
+    # Additionally, close all sessions.
+    close_all_sessions()
```

Comment:
I'd suggest flipping these - [from the SQLAlchemy docs](https://docs.sqlalchemy.org/en/13/core/connections.html?highlight=dispose#sqlalchemy.engine.Engine.dispose):

> This has the effect of fully closing all currently checked in database connections. Connections that are still checked out will not be closed, however they will no longer be associated with this Engine, so when they are closed individually, eventually the Pool which they are associated with will be garbage collected and they will be closed out fully, if not already closed on checkin.

Reply:
To clarify: 
```
  # Additionally, close all sessions to return connections to the pool
  close_all_sessions()
  # Close the database to clean up any connections
 if db:
        db.dispose()
        logger.info("Database connection disposed.")
```